server/src/main/java/org/apache/cassandra/sidecar/tasks/ClusterTopologyMonitor.java

```diff
@@ -0,0 +1,310 @@
+        }
+
+        public boolean equals(Object o)
+        {
+            if (this == o)
+            {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass())
+            {
+                return false;
+            }
+
+            DcLocalTopologyChangeEvent that = (DcLocalTopologyChangeEvent) o;
+            return Objects.equals(dc, that.dc) &&
+                   Objects.equals(prev, that.prev) &&
+                   curr.equals(that.curr);
+        }
+
+        public int hashCode()
+        {
+            return Objects.hash(dc, prev, curr);
+        }
+    }
+
+    /**
+     * Creates a cluster topology monitor with periodic task scheduling.
+     */
+    @Inject
+    public ClusterTopologyMonitor(Vertx vertx, TokenRingProvider tokenRingProvider, PeriodicTaskExecutor periodicTaskExecutor)
+    {
+        this(vertx, tokenRingProvider);
+        LOGGER.info("Starting Cluster Topology Monitor");
+        periodicTaskExecutor.schedule(this);
+        LOGGER.info("Cluster Topology Monitor started");
+    }
+
+    /**
+     * Creates a cluster topology monitor without scheduling.
+     */
+    public ClusterTopologyMonitor(Vertx vertx, TokenRingProvider tokenRingProvider)
+    {
+        this.vertx = vertx;
+        this.tokenRingProvider = tokenRingProvider;
+        try
+        {
+            vertx.eventBus().registerDefaultCodec(DcLocalTopologyChangeEvent.class, DcLocalTopologyChangeEventCodec.INSTANCE);
+        }
+        catch (IllegalStateException ise)
+        {
+            if (ise.getMessage().contains("Already a default codec registered"))
+            {
+                // TODO: do not register codec in the constructor when OSS; All codecs should be managed in a central repository
+                // ignore
+            }
+            else
+            {
+                throw ise;
+            }
+        }
+    }
+
+    /**
+     * Returns the delay between topology refresh cycles.
+     */
+    @Override
+    public DurationSpec delay()
+    {
+//        return MillisecondBoundConfiguration.parse(sidecarConfiguration.serviceConfiguration().refreshClusterTopology());
+        return MillisecondBoundConfiguration.parse("1000");
+    }
+
+    /**
+     * Executes the topology monitoring task.
+     */
+    @Override
+    public void execute(Promise<Void> promise)
+    {
+        try
+        {
+            refresh();
+        }
+        catch (Exception e)
+        {
+            LOGGER.error("Unexpected error in ClusterTopologyMonitor", e);
+        }
+        finally
+        {
+            promise.complete();
+        }
+    }
+
+    /**
+     * Refreshes topology information for all datacenters and publishes change events.
+     */
+    protected void refresh()
+    {
+        for (String dc : tokenRingProvider.dcs())
+        {
+            @Nullable final Map<String, List<TokenRange>> prev = perDcRanges.get(dc);
```

Comment:
We might not need `@nullable` for variable and  also noticed that other classes do not have final for local variables but this class seems to have final for local variable. Please remove them for consistency across this file.